roles/letsencrypt/tasks/dns-challenge-os.yml

```diff
@@ -0,0 +1,54 @@
+---
+- name: add new TXT record to DNS zone
+  openstack.cloud.recordset:
+    auth:
+      auth_url: "{{ dns_os_auth_url }}"
+      username: "{{ dns_user }}"
+      password: "{{ dns_password }}"
+      project_name: "{{ dns_os_project_name }}"
+      user_domain_name: "{{ dns_os_user_domain }}"
+    state: present
+    recordset_type: "txt"
+    records:
+      - "\"{{ challenge['challenge_data'][item]['dns-01']['resource_value'] }}\""
+    name: "_acme-challenge.{{ item | replace('*.', '') }}."
+    zone: "{{ domain.zone }}."
+  loop: "{{ domain.subject_alt_name }}"
+  register: records
+  when:
+    - domain.subject_alt_name is defined
+    - challenge['challenge_data'][item] is defined
+
+- name: Let the challenge be validated and retrieve the cert and intermediate certificate
+  acme_certificate:
+    account_key_src: "{{ account_key_path }}"
+    account_email: "{{ account_email }}"
+    csr: "{{ csr_path }}"
+    cert: "{{ cert_path }}"
+    fullchain: "{{ fullchain_path }}"
+    chain: "{{ intermediate_path }}"
+    challenge: dns-01
+    force: "{{ force_renewal | default(false) }}"
+    acme_directory: "{{ acme_directory }}"
+    acme_version: 2
+    terms_agreed: true
+    remaining_days: "{{ remaining_days }}"
+    data: "{{ challenge }}"
+
+- name: remove TXT records from DNS zone
+  openstack.cloud.recordset:
+    auth:
+      auth_url: "{{ dns_os_auth_url }}"
+      username: "{{ dns_user }}"
+      password: "{{ dns_password }}"
+      project_name: "{{ dns_os_project_name }}"
+      user_domain_name: "{{ dns_os_user_domain }}"
+    state: absent
+    name: "{{ item.name }}"
+    zone: "{{ item.zone_name }}"
+  loop: "{{ records | json_query('results[*].recordset') }}"
+  when:
```

Comment:
Why are these for when-conditions necessary? This seems very complicated.
Same goes for the regex - what does it do and why is it there?

Reply:
dcae7dd
Regex is there, because the loop is over "records", so "item.zone_name" is same as "item" of "domain.subject_alt_name" from first task (dns txt creation), with the slightly different, that "item.zone_name" ends with an unwanted dot.